### ISSUE TITLE
should a repo be a normal npub?

### ISSUE THREAD

**Issue by PeerRich** [NostrGit/NostrGit#4]
i wonder if a repo should just be an npub key like a normal nostr user

we would get:

- nip5
- zap a repo
- follow a repo (star)
- description
- profile photo

**Comment by cypherhoodlum** [comment]
I see where you're coming from, but I don't think we should reinvent repos. Nostrhub pages only track issues, PRs etc. which are nostr events. The repos themselves (code) are not on nostr.

**Comment by PeerRich** [comment]
i like nip5/repo_name

**Comment by PeerRich** [comment]
I agree @jbb55 recommended to try to use as much native nostr events as possible

**Comment by PeerRich** [comment]
> The repos themselves (code) are not on nostr. Correct me if I'm wrong.

i dont think thats possibe, most relays would reject such big payload

**Comment by melvincarvalho** [comment]
Yes, I came to the conclusion that repos benefit from having an npub (or nrepo).  In this way a repo or code base can participate in the nostr network.

My current design has a top level file `nostr.json` which will contain the pubkey, or optionally, bech32 version, `nrepo`

https://github.com/nip17/nips/blob/master/17.md#nostrjson

```JSON
{
    "pubkey": "abcd123...",
    "nrepo": "nrepo1ris1683fw6n2mvhl5h6dhqd8mqfv3wmxnz4qph83ua4dk4006ezsrt5c24"
}
```

Not sure if that's the best way to make a public key public.  Open to ideas!

**Comment by PeerRich** [comment]
agree that is a really fun idea to introduce nrepo!

**Comment by PeerRich** [comment]
i wonder if you could even do 

`git clone nrepo1ris1683fw6n2mvhl5h6dhqd8mqfv3wmxnz4qph83ua4dk4006ezsrt5c24` 🤔

**Comment by cypherhoodlum** [comment]
So would that make all holders of the private key code owners? The repo ownership hierarchy would be flatter than currently on GitHub for example. I wonder how the nuances of roles like Read, Triage, Write, Maintain, and Admin would work.

**Comment by melvincarvalho** [comment]
> i wonder if you could even do
> 
> `git clone nrepo1ris1683fw6n2mvhl5h6dhqd8mqfv3wmxnz4qph83ua4dk4006ezsrt5c24` thinking

yes, that's exactly what i have in mind, though perhaps building its own CLI as there's a few nuances, though likely all ends up in the same place

**Comment by melvincarvalho** [comment]
> So would that make all holders of the private key code owners? The repo ownership hierarchy would be flatter than currently on GitHub for example. I wonder how the nuances of roles like Read, Triage, Write, Maintain, and Admin would work.

When I started working on [NIP-17](http://nip17.com/) it lead to some confusion.

There's one group that want to make a better github over nostr.  Some prefer relays to hold the data, some prefer replays to hold references.  For that group each coder is the first class object, they have keys, and do stuff over relays.  Interact with git.  And patches and codes get merged to create a software development environment.

The thing I want is something different, though there is a big overlap.  I want every repo to be able to interact with the nostr network.  I want repo deployments to be able to interact with nostr.  And users to be able to interact with repos via nostr, and interact with repos via nostr.  Zap a repo, rate a repo, follow a repo, comment on a repo, add bounties.  Run code, that code interacts with nostr, can get zapped, can evolve etc.  Vote on how code evolves.  Have batteries for automated agents so they can keep running etc.

These are two different use cases.  One is a better github.  The other is git, code, and machines interacting with nostr.  So I think to avoid confusion different projects are going to have to go into a bit detail about which of those two use cases is being address, at any particular time.

In this issue giving a repo an npub so that it can become part of nostr is for me a really interesting step, but, as I found out, not everyone will wan that.

**Comment by PeerRich** [comment]
@cypherhoodlum i think account sharing will become a nostr-wide problem, requirement which is probably better solved on a protocol level.

**Comment by PeerRich** [comment]
> Zap a repo, rate a repo, follow a repo, comment on a repo, add bounties. Run code, that code interacts with nostr, can get zapped, can evolve etc. Vote on how code evolves. Have batteries for automated agents so they can keep running etc.

this is what I wanna achieve with this project too! 

zapping a repo could be a huge synergie!


![CleanShot 2023-03-06 at 13 44 42@2x](https://user-images.githubusercontent.com/8019099/223113871-35e03cc2-cff4-4ed0-a3e3-d1e0cc1d38ed.jpg)

**Comment by PeerRich** [comment]
also i think the synergy of using npub (or nrepo) for repos is that you can follow (star), store description, etc.

no need to merge an NIP for it -- already works

**Comment by PeerRich** [comment]
is someone close to @jb55? would love to hear his opinion given his work here: http://git.jb55.com/git-nostr-tools/file/README.txt.html

**Comment by PeerRich** [comment]
also links to repos could be as easy as:

nostrgit.com/nrepo1ris1683fw6n2mvhl5h6dhqd8mqfv3wmxnz4qph83ua4dk4006ezsrt5c24

**Comment by PeerRich** [comment]
and with nip5 maybe:

`git clone git@github.com:NostrGit/NostrGit.git`

will be

`git clone git@nostrgit.com:repo@organisation.com.git`

**Comment by cypherhoodlum** [comment]
> zapping a repo could be a huge synergy!

Also zapping individual PRs would incentivise FOSS contributions in all projects. I really like the idea of everything being events you can interact with via nostr.

**Comment by PeerRich** [comment]
i would love that!

**Comment by PeerRich** [comment]
i would love to zap you for this idea already!

**Comment by melvincarvalho** [comment]
> also i think the synergy of using npub (or nrepo) for repos is that you can follow (star), store description, etc.
> 
> no need to merge an NIP for it -- already works

Right now, npub and nrepo are just the same bech32 encoding of a public key.  But nrepo gives you a strong hint that we are dealing with a repo.  That in turn can help you choose which relays to query and which filters to apply.

To bech32 encode something you just add a prefix, encode it, and it adds a checksum.

What I have in mind for nrepo:
- paste in an input box and it brings up the whole history of the repo
- clone it with a CLI command and you get the latest code
- run an info command to get a history of the repo
- plug it into a continuous deployment tool and run the code
- put it in a live environment (e.g. docker) and it runs, but can also get live updates
- testing in local environments, live updates happen as you code
- more stuff

I chatted with fiatjaf about it, and he suggested also having a relay optional in the string.  So you could have nrepo1235..6?relays=wss...

In short it starts out as the pubkey of the repo, but can be extended when needs arise.

**Comment by cypherhoodlum** [comment]
> that way you could even interact with your normal nostr client (get notifications etc.)
It should be kept in mind that devs also need a space where to discuss technicals separately from non-devs. Right now the barrier to commenting is the github/lab account requirement. If that barrier doesn't exist, organising development can get convoluted very fast.

**Comment by PeerRich** [comment]
> organising development can get convoluted very fast.

interesting. very fair point.

exclusive relay maybe?

**Comment by cypherhoodlum** [comment]
I'm creating a summary of this discussion in documentation/vision.md. It can serve as a base to all major design desicions. I will cherry pick things I like, but let's see what kind of consensus forms over time, and update the document accordingly.

**Comment by PeerRich** [comment]
great idea

**Comment by ofer-elrom** [comment]
> In this issue giving a repo an npub so that it can become part of nostr is for me a really interesting step, but, as I found out, not everyone will want that.

I like very much this direction of git service suppliers and users interacting with them through Nostr.
IMO, censorship resistance to open source development is very important and can be achieved with NostrGit. 
A way to do this can be Nostr-authenticated developers who can use a Nostr-compatible git client to push to and pull from Nostr-authenticated git suppliers using Nostr relays as a communication layer.
For a more elaborate explanation, please see this google document: [docs.google.com/document/d/1z96XpTtrUvrFZYPjMzlT5JuE41HF6X5vI99lvImVCvU](https://docs.google.com/document/d/1z96XpTtrUvrFZYPjMzlT5JuE41HF6X5vI99lvImVCvU/)

**Comment by cypherhoodlum** [comment]
That's the plan. I think that's a very good summary of what we are building here. :+1: 
About the link: please don't use privacy-nightmare tools (google docs) when engaging in the project. You can try to break down complicated concepts into multiple issues and link them together with the # notation. That way complicated concepts morph into approachable chunks of todos, which turn into mergeable PRs.

**Comment by melvincarvalho** [comment]
Something I made quite quickly.  To run type:

> npx nrepo -p f4382214e53d53a291093baa9c13cea0aa44e4823d10c6ae362fbfaf4646a76a

Output:

```JSON
{
  "npub": "npub1k9lvh07g7md3llzejxykejqpg5hatqyrnkdf2a9s3rmzy5crvwtse5h9h3",
  "nrepo": "nrepo1k9lvh07g7md3llzejxykejqpg5hatqyrnkdf2a9s3rmzy5crvwtss0ahss"
}
```

which is based on noskey : https://github.com/melvincarvalho/noskey

I'll make some proper tooling over time, but it can a CLI to
- create new `nrepo` 
- clone an existing `nrepo` based on relay discovery
- publish `nrepo` to relays
- run an `nrepo` (think docker for nostr!)

I'll also do a simple doc explaining what `nrepo` is

**Comment by cypherhoodlum** [comment]
Fantastic. Please do write that doc so everyone can wrap their heads around your idea. What do you mean by 

> publish nrepo to relays

Previous discussion has been around using [GitTorrent](https://github.com/cjb/GitTorrent) to publish repos. We probably don't want to publish repos on relays, only references, to not flush out small relay operators with unreasonable scaling requirements for their relays. That being said, a working solution will probably override any other plans for it. Feel free to experiment.

I was thinking the CLI could work something like this:
```bash
# creates a normal git repository with pubkey and nrepo appended to a file
# runs git init under the hood && touch somerepo/.git/git-daemon-export-ok
$ nostrgit init
# publish the repo
# runs the gittorrentd command (gittorrent)
$ nostrgit serve
```
No idea if it could actually work, tho.

**Comment by melvincarvalho** [comment]
@cypherhoodlum that's cool.  The use case Im tackling first will publish the nrepo and a c (commit) tag with two entries

- commit hash
- git url [optional]

With these two things you can check out a repo, or move it from one git hosting service to another.  But I'll document this when I build out the CLI, this is just an advance idea of what I have planned.
